src/nifreeze/model/pet.py

```diff
@@ -31,60 +33,86 @@
+def _exec_fit(model, data, chunk=None, **kwargs):
+    return model.fit(data, **kwargs), chunk
 
-    __slots__ = (
-        "_t",
-        "_x",
-        "_xlim",
-        "_order",
-        "_n_ctrl",
-        "_datashape",
-        "_mask",
-        "_smooth_fwhm",
-        "_thresh_pct",
-    )
+
+def _exec_predict(model, chunk=None, **kwargs):
+    """Propagate model parameters and call predict."""
+    return np.squeeze(model.predict(**kwargs)), chunk
+
+
+class BasePETModel(BaseModel):
+    """Interface and default methods for PET models."""
+
+    __slots__ = {
+        "_data_mask": "A mask for the voxels that will be fitted and predicted",
+        "_x" : "",
+        "_xlim" : "",
+        "_smooth_fwhm" : "FWHM in mm over which to smooth",
+        "_thresh_pct" : "Thresholding percentile for the signal",
+        "_model_class": "Defining a model class",
+        "_modelargs": "Arguments acceptable by the underlying model",
+        "_models": "List with one or more (if parallel execution) model instances",
+    }
 
     def __init__(
         self,
-        dataset,
-        timepoints=None,
-        xlim=None,
-        n_ctrl=None,
-        order=3,
-        smooth_fwhm=10,
-        thresh_pct=20,
+        dataset: PET,
+        timepoints: list | np.ndarray = None,  ## Is there a way to use array-like
+        xlim: list | np.ndarray = None,
+        smooth_fwhm: float = 10.0,
+        thresh_pct : float = 20.0,
         **kwargs,
     ):
-        """
-        Create the B-Spline interpolating matrix.
+        """Initialization.
 
-        Parameters:
-        -----------
-        timepoints : :obj:`list`
+        Parameters
+        ----------
+        timepoints : :obj:`list` or :obj:`~np.ndarray`
             The timing (in sec) of each PET volume.
             E.g., ``[15.,   45.,   75.,  105.,  135.,  165.,  210.,  270.,  330.,
             420.,  540.,  750., 1050., 1350., 1650., 1950., 2250., 2550.]``
-
-        n_ctrl : :obj:`int`
-            Number of B-Spline control points. If `None`, then one control point every
-            six timepoints will be used. The less control points, the smoother is the
-            model.
-
+        xlim : .
+            .
+        smooth_fwhm : obj:`float`
+            FWHM in mm over which to smooth the signal.
+        thresh_pct : obj:`float`
+            Thresholding percentile for the signal.
         """
+
         super().__init__(dataset, **kwargs)
 
+        # Duck typing, instead of explicitly testing for PET type
+        if not hasattr(dataset, "total_duration"):
+            raise TypeError("Dataset MUST be a PET object.")
+
+        if not hasattr(dataset, "midframe"):
+            raise ValueError("Dataset MUST have a midframe.")
+
+        # ToDO
+        # Are the timepoints your "gradients" ??? If so, can they be computed
+        # from frame_time or frame_duration
+        # Or else frame_time and frame_duration ????
```

Comment:
@mnoergaard question for you. By your "gradients" I mean the parallel to the DWI cases, where the gradients are the independent variable where we will be willing to estimate the signal.

Reply:
Timepoints for PET are the gradients of DWI. We subsample the PET data into frames, and frames have a start time and an end time, and the values from the PET data are the uptake within this time period. Therefore, we use the middle timepoint (midframe) as timepoint.

Reply:
OK, understood. Thanks. Removing the comment and resolving.